src/Microsoft.OData.Core/ObjectPool.cs

```diff
@@ -0,0 +1,100 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ObjectPool.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// Manages a pool of reusable objects. An object retrieved from the pool should be returned when it's no longer being used so that it can be later recycled.
+    /// </summary>
+    /// <typeparam name="T">The type of objects to pool.</typeparam>
+    /// <remarks>This object pool is not thread-safe and not intended to be shared across concurrent requests.
+    /// Consider using the DefaultObjectPool in 8.x once we drop support for netstandard 1.1
+    /// </remarks>
+    internal class ObjectPool<T> where T : class
+    {
+        private readonly Func<T> objectGenerator;
+        private ObjectWrapper[] items;
+        private T firstItem;
+
+        // The pool size is equal to the level of nesting in the response.
+        // 32 is an arbitrary figure. Could be adjusted appropriately.
+        private const int POOL_SIZE = 8;
+
+        /// <summary>
+        /// To initialize the object pool.
+        /// </summary>
+        /// <param name="objectGenerator">Used to create an instance of a <typeparamref name="T"/>.</param>
+        public ObjectPool(Func<T> objectGenerator)
+        {
+            this.objectGenerator = objectGenerator ?? throw new ArgumentNullException(nameof(objectGenerator));
+        }
+
+        /// <summary>
+        /// Gets an object from the pool if one is available, otherwise creates one. The object will need to be returned to the pool when no longer in use.
+        /// </summary>
+        /// <returns>An object of type <typeparamref name="T"/> from the pool.</returns>
+        /// <remarks>It's the responsibility of the caller to clean up the object before using it.</remarks>
+        public T Get()
+        {
+            T item = firstItem;
+
+            if (item != null)
+            {
+                firstItem = default(T);

```

Comment:
Is this some kind of optimization for when we are in the same nesting level so that we don't have to iterate over the loop many times?

Reply:
Yes